# 📦 Installation et Configuration de l'Environnement Python pour ML/LLM

Bienvenue dans ce premier cours ! Avant de coder, il est essentiel de bien configurer votre environnement de travail. Ce notebook vous guidera pas à pas.

## 🎯 Objectifs
- Comprendre pourquoi utiliser **Miniconda** pour le Machine Learning
- Maîtriser les environnements virtuels
- Installer les packages essentiels pour le ML/LLM
- Vérifier que tout fonctionne correctement

---
## 1️⃣ Installation de Python avec Miniconda

### Pourquoi Miniconda plutôt qu'Anaconda ?

| Miniconda | Anaconda |
|-----------|----------|
| ~80 MB (léger) | ~3 GB (lourd) |
| Installe seulement le nécessaire | 8000+ packages pré-installés |
| Plus rapide à installer | Beaucoup de packages inutiles |
| **✅ Recommandé** | Trop volumineux |

**Miniconda** inclut :
- **Python**
- **Conda** : gestionnaire de packages et d'environnements
- C'est tout ! Tu installes ce dont tu as besoin, rien de plus.

---

### 📥 Installation sur Mac Apple Silicon (M1/M2/M3/M4)

1. Va sur : [https://www.anaconda.com/download](https://www.anaconda.com/download)
2. Dans la section **Miniconda** (à droite), clique sur :

   **→ 64-Bit (Apple silicon) Graphical Installer**

3. Ouvre le fichier `.pkg` téléchargé
4. Suis l'installation (clique "Continuer" plusieurs fois)
5. **Ferme et rouvre ton Terminal** (important !)

---

### 📥 Installation sur Windows

1. Va sur : [https://www.anaconda.com/download](https://www.anaconda.com/download)
2. Dans la section **Miniconda**, télécharge **64-Bit (Windows)**
3. Lance l'installateur `.exe`
4. **Important** : Coche "Add Miniconda to PATH" pendant l'installation
5. Utilise "Anaconda Prompt" pour les commandes conda

---

### ✅ Vérification de l'installation

Ouvre un **nouveau** Terminal (important : nouveau !) et tape :

```bash
conda --version
```

Tu devrais voir quelque chose comme `conda 24.x.x`

In [2]:
# Dans un terminal, exécutez ces commandes (PAS dans Jupyter pour l'instant)
# conda --version
# python --version

# Dans Jupyter, vous pouvez vérifier la version de Python comme ceci :
import sys
print(f"Version de Python : {sys.version}")
print(f"Chemin de l'exécutable Python : {sys.executable}")

Version de Python : 3.10.16 (main, Nov 15 2025, 19:04:17) [Clang 17.0.0 (clang-1700.3.19.1)]
Chemin de l'exécutable Python : /Users/sanz/.pyenv/versions/3.10.16/bin/python


**Résultat attendu :** Vous devriez voir quelque chose comme `Python 3.11.x` ou `Python 3.10.x`

---
## 2️⃣ Qu'est-ce qu'un Environnement Virtuel ?

### 🤔 Le problème

Imaginez que vous travaillez sur deux projets :
- **Projet A** : nécessite TensorFlow 2.10
- **Projet B** : nécessite TensorFlow 2.15

Si vous installez les deux versions au même endroit, **conflit garanti** ! 💥

### 💡 La solution : Environnements virtuels

Un environnement virtuel est comme une **boîte isolée** qui contient :
- Une version spécifique de Python
- Un ensemble de packages spécifiques
- Totalement indépendant des autres environnements

```
Votre ordinateur
├── Environnement de base (base)
├── Environnement "llm" (Python 3.11, PyTorch 2.0, transformers)
├── Environnement "data-science" (Python 3.10, TensorFlow 2.10)
└── Environnement "web" (Python 3.9, Django)
```

### ✅ Avantages
- Pas de conflits entre projets
- Reproductibilité : facile de recréer l'environnement
- Propreté : supprimez un projet, supprimez son environnement
- Flexibilité : testez différentes versions sans risque

---
## 3️⃣ Créer un Environnement avec Conda

### 📝 Commande de base

Ouvrez un terminal et exécutez :

```bash
conda create -n llm python=3.11
```

**Décryptage de la commande :**
- `conda create` : crée un nouvel environnement
- `-n llm` : nomme l'environnement "llm" (vous pouvez choisir n'importe quel nom)
- `python=3.11` : installe Python version 3.11

### 🔄 Options utiles

```bash
# Créer avec des packages directement
conda create -n llm python=3.11 numpy pandas jupyter

# Créer avec une version spécifique de package
conda create -n llm python=3.11 numpy=1.24

# Créer depuis un fichier de configuration
conda create -n llm --file requirements.txt
```

### 📋 Commandes utiles pour gérer les environnements

```bash
# Lister tous les environnements
conda env list

# Supprimer un environnement
conda remove -n llm --all

# Cloner un environnement
conda create -n llm-backup --clone llm

# Exporter un environnement (pour le partager)
conda env export > environment.yml

# Créer depuis un export
conda env create -f environment.yml
```

---
## 4️⃣ Activer et Désactiver un Environnement

### 🟢 Activer un environnement

**Windows :**
```bash
conda activate llm
```

**Mac/Linux :**
```bash
conda activate llm
```

Votre terminal devrait maintenant afficher `(llm)` au début de la ligne, indiquant que l'environnement est actif.

### 🔴 Désactiver un environnement

```bash
conda deactivate
```

Vous reviendrez à l'environnement de base `(base)`.

### ⚠️ Important

- **Toujours activer l'environnement** avant d'installer des packages ou de lancer Jupyter
- Si Jupyter ne voit pas vos packages, vérifiez que vous êtes dans le bon environnement
- L'activation est **temporaire** : quittez le terminal, il faut réactiver

In [3]:
# Vérifier dans quel environnement vous êtes actuellement
import os
import sys

print(f"Environnement actif : {os.environ.get('CONDA_DEFAULT_ENV', 'Non détecté')}")
print(f"Chemin Python : {sys.executable}")

Environnement actif : Non détecté
Chemin Python : /Users/sanz/.pyenv/versions/3.10.16/bin/python


---
## 5️⃣ Installer des Packages : pip vs conda

### 🔄 Deux gestionnaires de packages

| Caractéristique | conda | pip |
|----------------|-------|-----|
| Source | Anaconda repository | Python Package Index (PyPI) |
| Langages | Python + R + C++ | Python uniquement |
| Dépendances | Gère tout (y compris non-Python) | Python uniquement |
| Vitesse | Plus lent | Plus rapide |
| Disponibilité | Moins de packages | Plus de packages |

### 📦 Installation avec conda

```bash
# Installation basique
conda install numpy

# Version spécifique
conda install numpy=1.24.0

# Plusieurs packages à la fois
conda install numpy pandas matplotlib

# Depuis un canal spécifique
conda install -c conda-forge transformers
```

### 📦 Installation avec pip

```bash
# Installation basique
pip install torch

# Version spécifique
pip install torch==2.0.0

# Plusieurs packages
pip install torch transformers datasets

# Depuis un fichier requirements.txt
pip install -r requirements.txt

# Mettre à jour un package
pip install --upgrade torch
```

### 🎯 Règle d'or

1. **Privilégiez conda** pour les packages scientifiques (numpy, scipy, pandas)
2. **Utilisez pip** si le package n'existe pas dans conda
3. **Évitez de mélanger** conda et pip autant que possible
4. Si vous devez mélanger : **conda d'abord, pip ensuite**

### 📋 Commandes utiles

```bash
# Lister les packages installés
conda list
pip list

# Rechercher un package
conda search numpy
pip search numpy  # (désactivé depuis 2021, utilisez pypi.org)

# Désinstaller un package
conda remove numpy
pip uninstall numpy

# Mettre à jour tous les packages
conda update --all
pip list --outdated  # liste les packages à mettre à jour
```

---
## 6️⃣ Packages Essentiels pour ML/LLM

Voici les packages à installer dans votre environnement `llm`. **Activez d'abord votre environnement !**

```bash
conda activate llm
```

### 📊 Data Science de base

```bash
conda install numpy pandas matplotlib seaborn
```

- **numpy** : calculs numériques, matrices, algèbre linéaire
- **pandas** : manipulation de données tabulaires (DataFrame)
- **matplotlib** : visualisation de données (graphiques)
- **seaborn** : visualisation statistique avancée

### 🤖 Machine Learning classique

```bash
conda install scikit-learn
```

- **scikit-learn** : algorithmes ML (classification, régression, clustering)

### 🧠 Deep Learning avec PyTorch

**Pour Mac Apple Silicon (M1/M2/M3/M4) :**
```bash
# Utilise automatiquement le GPU Metal (MPS)
pip install torch torchvision torchaudio
```

**Pour Windows/Linux CPU :**
```bash
conda install pytorch torchvision torchaudio cpuonly -c pytorch
```

**Pour Windows/Linux avec GPU NVIDIA :**
```bash
conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia
```

- **pytorch** : framework de deep learning
- **torchvision** : utilitaires pour la vision par ordinateur
- **torchaudio** : utilitaires pour l'audio

> 💡 **Note Mac M4** : PyTorch supporte le GPU Apple Silicon via MPS (Metal Performance Shaders). C'est automatique !

### 🤗 Hugging Face pour les LLM

```bash
pip install transformers datasets accelerate
```

- **transformers** : modèles pré-entraînés (GPT, BERT, LLaMA, etc.)
- **datasets** : accès à des milliers de datasets
- **accelerate** : entraînement distribué simplifié

### 📓 Jupyter et outils de développement

```bash
conda install jupyter ipykernel notebook
```

### 🎁 Packages bonus utiles

```bash
pip install tqdm pillow python-dotenv
```

---

### 📦 Installation complète en une commande

**Pour Mac Apple Silicon (M1/M2/M3/M4) :**
```bash
# 1. Créer l'environnement
conda create -n llm python=3.11 numpy pandas matplotlib seaborn scikit-learn jupyter ipykernel notebook requests -y

# 2. Activer
conda activate llm

# 3. Installer PyTorch (avec support GPU Metal/MPS)
pip install torch torchvision torchaudio

# 4. Installer Hugging Face et bonus
pip install transformers datasets accelerate tqdm pillow python-dotenv

# 5. Enregistrer dans Jupyter
python -m ipykernel install --user --name=llm --display-name="Python (LLM)"
```

**Pour Windows/Linux :**
```bash
# 1. Créer l'environnement
conda create -n llm python=3.11 numpy pandas matplotlib seaborn scikit-learn jupyter ipykernel notebook requests -y

# 2. Activer
conda activate llm

# 3. Installer PyTorch (CPU)
conda install pytorch torchvision torchaudio cpuonly -c pytorch -y

# 4. Installer Hugging Face et bonus
pip install transformers datasets accelerate tqdm pillow python-dotenv

# 5. Enregistrer dans Jupyter
python -m ipykernel install --user --name=llm --display-name="Python (LLM)"
```

---
## 7️⃣ Vérification des Installations

Exécutez les cellules suivantes pour vérifier que tout est bien installé.

In [4]:
# Vérification de l'environnement Python
import sys
print(f"Version de Python : {sys.version}")
print(f"Exécutable : {sys.executable}")

Version de Python : 3.10.16 (main, Nov 15 2025, 19:04:17) [Clang 17.0.0 (clang-1700.3.19.1)]
Exécutable : /Users/sanz/.pyenv/versions/3.10.16/bin/python


In [5]:
# Vérification des packages Data Science
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

print(f"✅ NumPy version : {np.__version__}")
print(f"✅ Pandas version : {pd.__version__}")
print(f"✅ Matplotlib version : {plt.matplotlib.__version__}")

ModuleNotFoundError: No module named 'numpy'

In [ ]:
# Vérification de scikit-learn
import sklearn
print(f"✅ Scikit-learn version : {sklearn.__version__}")

In [ ]:
# Vérification de PyTorch
import torch
print(f"✅ PyTorch version : {torch.__version__}")

# Vérification GPU
print(f"\n🎮 Accélération GPU :")
print(f"  CUDA (NVIDIA) : {torch.cuda.is_available()}")
print(f"  MPS (Apple Silicon) : {torch.backends.mps.is_available()}")

if torch.cuda.is_available():
    print(f"  → Device CUDA : {torch.cuda.get_device_name(0)}")
elif torch.backends.mps.is_available():
    print(f"  → Apple Silicon GPU disponible ! 🍎")
    print(f"    Ton Mac M4 peut accélérer les calculs PyTorch !")
else:
    print("  → Mode CPU uniquement")

In [ ]:
# Vérification de Transformers (Hugging Face)
import transformers
print(f"✅ Transformers version : {transformers.__version__}")

In [ ]:
# Test rapide : créer un petit graphique
import numpy as np
import matplotlib.pyplot as plt

x = np.linspace(0, 10, 100)
y = np.sin(x)

plt.figure(figsize=(8, 4))
plt.plot(x, y, label='sin(x)', color='blue', linewidth=2)
plt.xlabel('x')
plt.ylabel('y')
plt.title('Test de Matplotlib : fonction sinus')
plt.legend()
plt.grid(True, alpha=0.3)
plt.show()

print("✅ Si vous voyez un graphique ci-dessus, Matplotlib fonctionne !")

In [ ]:
# Test rapide : créer un DataFrame pandas
import pandas as pd

data = {
    'Nom': ['Alice', 'Bob', 'Charlie'],
    'Age': [25, 30, 35],
    'Ville': ['Paris', 'Lyon', 'Marseille']
}

df = pd.DataFrame(data)
print("✅ Test de Pandas - DataFrame créé :")
print(df)

In [ ]:
# Test rapide : opération avec PyTorch
import torch

# Créer un tenseur
tensor = torch.tensor([[1, 2], [3, 4]], dtype=torch.float32)
print("Tenseur original :")
print(tensor)

# Opération simple
result = tensor * 2 + 1
print("\nAprès multiplication par 2 et addition de 1 :")
print(result)

print("\n✅ PyTorch fonctionne correctement !")

### 🎉 Résumé des vérifications

Si toutes les cellules ci-dessus s'exécutent sans erreur, **félicitations !** Votre environnement est prêt pour le Machine Learning et les LLM.

**Checklist :**
- ✅ Python installé et version correcte
- ✅ NumPy, Pandas, Matplotlib fonctionnent
- ✅ Scikit-learn installé
- ✅ PyTorch installé (CPU ou GPU)
- ✅ Transformers (Hugging Face) installé
- ✅ Jupyter fonctionne (vous lisez ce notebook !)

---
## 8️⃣ Résolution de Problèmes Courants

### ❌ Problème 1 : "conda n'est pas reconnu"

**Symptôme :** En tapant `conda` dans le terminal, tu obtiens une erreur.

**Solutions Mac :**
1. Ferme et rouvre le Terminal (important !)
2. Si ça ne marche toujours pas, tape :
```bash
source ~/miniconda3/bin/activate
conda init zsh
```
3. Ferme et rouvre le Terminal

**Solutions Windows :**
1. Utilise "Anaconda Prompt" au lieu du terminal standard
2. Ou réinstalle Miniconda en cochant "Add to PATH"

---

### ❌ Problème 2 : "ModuleNotFoundError: No module named 'xxx'"

**Symptôme :** Python ne trouve pas un package que tu as installé.

**Causes possibles :**
1. Package installé dans un autre environnement
2. Jupyter utilise un autre kernel

**Solutions :**
```bash
# Vérifier l'environnement actif
conda env list  # L'environnement actif a une étoile *

# Activer le bon environnement
conda activate llm

# Installer le package
conda install nom-du-package

# Réenregistrer le kernel Jupyter
python -m ipykernel install --user --name=llm
```

---

### ❌ Problème 3 : Installation PyTorch échoue

**Pour Mac Apple Silicon :**
```bash
# Méthode recommandée
pip install torch torchvision torchaudio

# Si ça ne marche pas, essaye :
pip install --upgrade pip
pip install torch torchvision torchaudio
```

**Pour Windows/Linux :**
```bash
# Utiliser pip au lieu de conda
pip install torch torchvision torchaudio
```

---

### ❌ Problème 4 : Jupyter ne trouve pas l'environnement

**Solutions :**
```bash
# Activer l'environnement
conda activate llm

# Réinstaller ipykernel
pip install ipykernel

# Enregistrer le kernel
python -m ipykernel install --user --name=llm --display-name="Python (LLM)"

# Lister les kernels disponibles
jupyter kernelspec list
```

---

### ❌ Problème 5 : MPS (Apple Silicon GPU) non détecté

**Symptôme :** `torch.backends.mps.is_available()` retourne `False`

**Solutions :**
```bash
# Vérifier la version de macOS (12.3+ requis pour MPS)
sw_vers

# Réinstaller PyTorch
pip uninstall torch torchvision torchaudio
pip install torch torchvision torchaudio
```

---

### 🆘 Commandes de diagnostic

Exécutez ces commandes pour diagnostiquer les problèmes :

In [ ]:
# Diagnostic complet de l'environnement
import sys
import os

print("=" * 50)
print("DIAGNOSTIC DE L'ENVIRONNEMENT")
print("=" * 50)

print(f"\n📍 Version Python : {sys.version}")
print(f"📍 Exécutable Python : {sys.executable}")
print(f"📍 Environnement actif : {os.environ.get('CONDA_DEFAULT_ENV', 'Non détecté')}")

print("\n📦 Packages installés :")
packages = ['numpy', 'pandas', 'matplotlib', 'sklearn', 'torch', 'transformers']

for pkg in packages:
    try:
        module = __import__(pkg)
        version = getattr(module, '__version__', 'Version inconnue')
        print(f"  ✅ {pkg:15} : {version}")
    except ImportError:
        print(f"  ❌ {pkg:15} : NON INSTALLÉ")

print("\n🎮 GPU Info :")
try:
    import torch
    print(f"  CUDA (NVIDIA) : {torch.cuda.is_available()}")
    print(f"  MPS (Apple Silicon) : {torch.backends.mps.is_available()}")
    if torch.cuda.is_available():
        print(f"  → Device CUDA : {torch.cuda.get_device_name(0)}")
    elif torch.backends.mps.is_available():
        print(f"  → Apple Silicon GPU disponible ! 🍎")
except Exception as e:
    print(f"  PyTorch non installé ou problème : {e}")

print("\n" + "=" * 50)

---
## 📚 Ressources Supplémentaires

### Documentation Officielle
- [Miniconda Documentation](https://docs.anaconda.com/miniconda/)
- [Conda Cheat Sheet](https://docs.conda.io/projects/conda/en/latest/user-guide/cheatsheet.html)
- [PyTorch Installation](https://pytorch.org/get-started/locally/)
- [PyTorch sur Apple Silicon](https://developer.apple.com/metal/pytorch/)
- [Hugging Face Transformers](https://huggingface.co/docs/transformers/installation)

### Tutoriels
- [Real Python - Virtual Environments](https://realpython.com/python-virtual-environments-a-primer/)
- [Jupyter Notebook Tutorial](https://jupyter-notebook.readthedocs.io/)

### Communautés
- [Stack Overflow](https://stackoverflow.com/questions/tagged/python)
- [Reddit r/learnpython](https://www.reddit.com/r/learnpython/)
- [Hugging Face Forum](https://discuss.huggingface.co/)

---
## ✅ Checklist Finale

Avant de passer au cours suivant, assure-toi que :

- [ ] Miniconda est installé et fonctionne (`conda --version`)
- [ ] Tu comprends le concept d'environnement virtuel
- [ ] Tu as créé l'environnement `llm`
- [ ] Tu sais activer/désactiver un environnement
- [ ] Tu comprends la différence entre conda et pip
- [ ] Tous les packages essentiels sont installés et testés
- [ ] Jupyter Notebook fonctionne avec le bon kernel
- [ ] Tu sais où trouver de l'aide en cas de problème

---

## 🎯 Prochaine Étape

Maintenant que ton environnement est prêt, tu peux passer aux **cours de Mathématiques** dans `Phase_1_Mathematiques/` !

On va apprendre :
- Les bases des maths (puissances, racines, variables)
- Les fonctions et leur représentation graphique
- La notation Sigma (Σ) - essentielle pour le ML !

**Bon apprentissage ! 🚀**